File: NtgeCore-Android/src/main/java/com/dimension/ntge/message/Encryptor.kt
package com.dimension.ntge.message

import com.dimension.ntge.Ntge
import com.dimension.ntge.ed25519.Ed25519PrivateKey
import com.dimension.ntge.x25519.X25519PublicKey

class Encryptor internal constructor(
        internal val ptr: Long,
        private val keys_ptr: Long
) : AutoCloseable {

    companion object {
        fun new(vararg x25519PublicKey: X25519PublicKey): Encryptor {
            val keyPtr = Ntge.newArrayForX25519PublicKey()
            x25519PublicKey.forEach {
                Ntge.pushArrayX25519PublicKey(keyPtr, it.ptr)
            }
            return Encryptor(Ntge.newMessageEncryptor(keyPtr), keyPtr)
        }
    }

    fun encryptPlaintext(input: String, signatureKey: Ed25519PrivateKey? = null): Message {
        return Ntge.encryptPlaintext(input, ptr, signatureKey?.ptr
                ?: 0).let {
            Message(it)
        }
    }

    override fun close() {
        Ntge.destroyMessageEncryptor(ptr)
        Ntge.destroyArrayX25519PublicKey(keys_ptr)
    }

}

File: NtgeCore.Net/src/Message/Encryptor.cs
using System;
using System.Text;
using NtgeCore.Net.Ed25519;
using NtgeCore.Net.X25519;

namespace NtgeCore.Net.Message
{
    public class Encryptor : RustObject
    {
        private IntPtr _keysPtr;
        internal Encryptor(IntPtr ptr, IntPtr keysPtr) : base(ptr)
        {
            _keysPtr = keysPtr;
        }

        public static Encryptor New(params X25519PublicKey[] publicKey)
        {
            var keysPtr = Native.newArrayForX25519PublicKey();
            foreach (var item in publicKey)
            {
                Native.pushArrayX25519PublicKey(keysPtr, item.Ptr);
            }
            return new Encryptor(Native.newMessageEncryptor(keysPtr), keysPtr);
        }

        public NtgeMessage EncryptPlaintext(string text, Ed25519PrivateKey? signatureKey = null)
        {
            return new NtgeMessage(Native.encryptPlaintext(Encoding.UTF8.GetBytes(text), Ptr, signatureKey?.Ptr ?? IntPtr.Zero));
        }

        public NtgeMessage EncryptPlaintextWithExtra(string text, string extra, Ed25519PrivateKey? signatureKey = null)
        {
            return new NtgeMessage(Native.encryptPlaintextWithExtra(Ptr, Encoding.UTF8.GetBytes(text), Encoding.UTF8.GetBytes(extra) , signatureKey?.Ptr ?? IntPtr.Zero));
        }

        public override void Dispose()
        {
            Native.destroyMessageEncryptor(Ptr);
            Native.destroyArrayX25519PublicKey(_keysPtr);
        }
    }
}